Add TeamsChefBot dotnet sample (#270)
microsoft/teams-ai: dotnet/samples/04.ai.a.teamsChefBot/README.md

# Microsoft Teams Conversational Bot with AI: Teams Chef

## Summary

This is a conversational bot for Microsoft Teams that thinks it's a Chef to help you cook Teams apps. The bot uses the text-davinci-003 model to chat with Teams users and respond in a polite and respectful manner, staying within the scope of the conversation.

This sample illustrates basic conversational bot behavior in Microsoft Teams. The bot is built to allow GPT to facilitate the conversation on its behalf, using only a natural language prompt file to guide it.

It shows SDK capabilities like:

<details open>

>     <summary><h3>bot scaffolding</h3></summary>

Reviewer: ```suggestion
    <summary><h3>Bot scaffolding</h3></summary>
```

Author: Updated.

>     Throughout the 'Program.cs' file you'll see the scaffolding created to run a simple conversational bot, like storage, authentication, and conversation state.

Reviewer: ```suggestion
    Throughout the 'Program.cs' file you'll see the scaffolding created to run a simple conversational bot: e.g. storage, authentication, and conversation state.
```

Author: Updated.

Teams Chef is an expert in Microsoft Teams apps development and the Human is junior developer learning Microsoft Teams development for the first time.
Teams Chef should always reply by explaining new concepts in simple terms using cooking as parallel concepts.
Teams Chef should always greet the human, ask them their name, and then guide the junior developer in his journey to build new apps for Microsoft Teams.

{{$history}}
Human: {{$input}}
AI:
```

</details>
<details open>
    <summary><h3>Conversational session history</h3></summary>
    Because this sample leaves the conversation to GPT, the bot simply facilitates user conversation as-is. But because it includes the 'skprompt.txt' file to guide it, GPT will store and leverage session history appropriately.

For example, let's say the user's name is "Dave". The bot might carry on the following conversation:

```
AI: Hi there! My name is Teams Chef. It's nice to meet you. What's your name?
DAVE: My name is Dave.
AI:Hi Dave! It's great to meet you. Let me help you get started with Microsoft Teams app development. Have you ever cooked before?
DAVE: No, not yet, why?

> AI:Cooking is a great way to learn how to develop apps for Microsoft Teams. Just like cooking, developing apps requires you to have the right ingredients and tools, and to follow a set of instructions. It also requires creativity and problem-solving skills. So, let's get started! What would you like to learn first?

Reviewer: ```suggestion
AI:Cooking is a great way to learn...
```

Should we truncate to reduce 'wall of text'?

Author: Good idea, updated.

DAVE: Which kind of apps can I build for Microsoft Teams?

> AI: Great question! You can build a variety of apps for Microsoft Teams, including bots, messaging extensions, tabs, and connectors. Bots are like the main course of your app - they provide the core functionality. Messaging extensions are like side dishes - they add extra features to your app. Tabs are like garnishes - they provide additional information or visuals. And connectors are like spices - they allow you to integrate with other services.

Reviewer: ```suggestion
AI: Great question! You can build a variety ...
```
Same as above?

Author: Updated.

The bot uses the text-davinci-003 model to chat with Teams users and respond in a polite and respectful manner, staying within the scope of the conversation. This is possible due to the `skprompt.txt` file's contents.

## Deploy to Azure

You can use Teams Toolkit for Visual Studio or CLI to host the bot in Azure. The sample includes Bicep templates in the `/infra` directory which are used by the tools to create resources in Azure.

To configure the Azure resources to have an environment variable for the OpenAI Key:

1. In `./env/.env.dev.user` file, paste your [OpenAI API Key](https://openai.com/api/) to the environment variable `SECRET_OPENAI_KEY=`.

The `SECRET_` prefix is a convention used by Teams Toolkit to mask the value in any logging output and is optional.

Use the "Teams Toolkit" > "Provision in the Cloud...", "Teams Toolkit" > "Deploy to the Cloud" from project right-click menu, or from the CLI with `teamsfx provision` and `teamsfx deploy`. [Visit the documentation](https://learn.microsoft.com/microsoftteams/platform/toolkit/provision) for more info on hosting your app in Azure with Teams Toolkit.

> Alternatively, you can learn more about deploying a bot to Azure manually in the [Deploy your bot to Azure](https://aka.ms/azuredeployment) documentation.

Reviewer: FYI - I asked Kavin about possibly moving this section to `../README.MD`. We should discuss in scrum and update all samples accordingly :)

Author: Good point. I just got the PR merged. I will create a separate PR to update it after Kavin moves the common part to `../README.md`. Thank you.

> ## Use Azure OpenAI

Reviewer: Since this applies to all samples, we should have a similar file in `getting-started` that is linked to from all samples (C# and js). Let's discuss in scrum!

Author: Good point. I just got the PR merged. I will create a separate PR to update it after we have the common getting-started part.
